src/plugins/data_views/server/rest_api_routes/internal/has_es_data.ts

```diff
@@ -7,34 +7,122 @@
-  const core = await ctx.core;
-  const elasticsearchClient = core.elasticsearch.client.asCurrentUser;
-  const response = await elasticsearchClient.indices.resolveCluster({
-    name: patterns.concat(crossClusterPatterns),
-    allow_no_indices: true,
-    ignore_unavailable: true,
-  });
+export const createHandler =
+  (parentLogger: Logger, hasEsDataTimeout: number): Handler =>
+  async (ctx, _, res) => {
+    const logger = parentLogger.get('hasEsData');
+    const core = await ctx.core;
+    const elasticsearchClient = core.elasticsearch.client.asCurrentUser;
+    const commonParams: Omit<HasEsDataParams, 'matchPatterns' | 'timeoutReason'> = {
+      elasticsearchClient,
+      logger,
+      res,
+      hasEsDataTimeout,
+    };
 
-  const hasEsData = !!Object.values(response).find((cluster) => cluster.matching_indices);
+    const localDataResponse = await hasEsData({
+      ...commonParams,
+      matchPatterns: patterns,
+      timeoutReason: HasEsDataFailureReason.localDataTimeout,
+    });
 
-  return res.ok({ body: { hasEsData } });
+    if (localDataResponse) {
+      return localDataResponse;
+    }
+
+    const remoteDataResponse = await hasEsData({
+      ...commonParams,
+      matchPatterns: crossClusterPatterns,
+      timeoutReason: HasEsDataFailureReason.remoteDataTimeout,
+    });
+
+    if (remoteDataResponse) {
+      return remoteDataResponse;
+    }
+
+    return res.ok({ body: { hasEsData: false } });
+  };
+
+interface HasEsDataParams {
+  elasticsearchClient: ElasticsearchClient;
+  logger: Logger;
+  res: KibanaResponseFactory;
+  matchPatterns: string[];
+  hasEsDataTimeout: number;
+  timeoutReason: HasEsDataFailureReason;
+}
+
+const timeoutMessage = 'Timeout while checking for Elasticsearch data';
+const errorMessage = 'Error while checking for Elasticsearch data';
+
+const hasEsData = async ({
+  elasticsearchClient,
+  logger,
+  res,
+  matchPatterns,
+  hasEsDataTimeout,
+  timeoutReason,
+}: HasEsDataParams) => {
+  try {
+    const response = await elasticsearchClient.indices.resolveCluster(
+      {
+        name: matchPatterns,
+        allow_no_indices: true,
+        ignore_unavailable: true,
+      },
+      { requestTimeout: hasEsDataTimeout === 0 ? undefined : hasEsDataTimeout }
+    );
+
+    const hasData = Object.values(response).some((cluster) => cluster.matching_indices);
+
+    if (hasData) {
+      return res.ok({ body: { hasEsData: true } });
+    }
+  } catch (e) {
+    if (e.name === 'TimeoutError') {
+      const warningMessage =
+        `${timeoutMessage}: ${timeoutReason}. Current timeout value is ${hasEsDataTimeout}ms. ` +
+        `Use "data_views.hasEsDataTimeout" in kibana.yml to change it, or set to 0 to disable timeouts.`;
+
+      logger.warn(warningMessage);
+
+      return res.badRequest({
+        body: {
+          message: timeoutMessage,
+          attributes: { failureReason: timeoutReason },
+        },
+      });
```

Comment:
I'm not sure a "bad request" response makes sense here... The client didn't send anything wrong. Maybe we an do a custom error with a [408 status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408)?

```suggestion
      return res.customError({
        body: {
          statusCode: 408,
          message: timeoutMessage,
          attributes: { failureReason: timeoutReason },
        },
      });
```

Reply:
I agree, 400 isn't really appropriate for this. I originally looked at 408 too, but my understanding is that it's used when a server times out waiting on a request from the client, not when it times out trying to return a response. After reading into it a bit more, I feel like [504 Gateway Timeout](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504) might be most appropriate for this case, so I updated it here: https://github.com/elastic/kibana/pull/200476/commits/f130736e12d6c1fe546b13f1a03f0b75473a10b3.

I was hoping to avoid a generic 500 since it may be misleading and look like a Kibana server failure, but we could instead just go with that if 504 doesn't seem good either.